src/charm.py

```diff
@@ -196,401 +221,149 @@ def __init__(self, *args):
-            return
-
-        if not self._peers:
-            self.unit.status = WaitingStatus("Waiting for peer relation")
-            event.defer()
-            return
-
-        if openfga_info := self._get_openfga_store_info():
-            self._create_openfga_model(openfga_info)
-
-    def _get_openfga_store_info(self) -> Optional[OpenfgaProviderAppData]:
-        openfga_info = self.openfga.get_store_info()
-        if not openfga_info or not openfga_info.store_id:
-            logger.info("No openfga store info available")
-            return None
-
-        return openfga_info
-
-    def _handle_status_update_config(self, event: HookEvent) -> None:  # noqa: C901
-        if not self._container.can_connect():
-            event.defer()
-            logger.info("Cannot connect to admin-ui container. Deferring the event.")
-            self.unit.status = WaitingStatus("Waiting to connect to admin-ui container")
-            return
-
-        if not self._peers:
-            self.unit.status = WaitingStatus("Waiting for peer relation")
-            event.defer()
-            return
-
-        if not self.model.relations[KRATOS_INFO_INTEGRATION_NAME]:
-            self.unit.status = BlockedStatus("Missing required relation with kratos")
-            return
-
-        if not self.model.relations[HYDRA_ENDPOINTS_INTEGRATION_NAME]:
-            self.unit.status = BlockedStatus("Missing required relation with hydra")
-            return
+    @wait_when(container_not_connected)
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        self._workload_service.push_ca_certs(event.ca)
+        self._handle_status_update_config(event)
 
-        if not self.model.relations[OPENFGA_INTEGRATION_NAME]:
-            self.unit.status = BlockedStatus("Missing required relation with openfga")
-            return
+    @wait_when(container_not_connected)
+    def _on_certificate_removed(self, event: CertificateRemovedEvent) -> None:
+        self._workload_service.remove_ca_certs()
+        self._handle_status_update_config(event)
 
-        if (
-            self.model.relations[OAUTH_INTEGRATION_NAME]
-            and not self.model.relations[INGRESS_INTEGRATION_NAME]
+    @wait_when(
+        container_not_connected,
+        integration_not_exists(PEER_INTEGRATION_NAME),
+    )
+    @block_when(
+        integration_not_exists(KRATOS_INFO_INTEGRATION_NAME),
+        integration_not_exists(HYDRA_ENDPOINTS_INTEGRATION_NAME),
+        integration_not_exists(OPENFGA_INTEGRATION_NAME),
+        integration_not_exists(INGRESS_INTEGRATION_NAME),
+    )
+    def _handle_status_update_config(self, event: EventBase) -> None:
+        if self.oauth_integration.is_ready() and (
+            not self.model.relations[CERTIFICATE_TRANSFER_INTEGRATION_NAME]
         ):
-            self.unit.status = BlockedStatus("Ingress relation is required for oauth")
-            return
-
-        if not self._validate_cert_trust():
```

Comment:
I chose the EAFP instead of LBYL because it's simpler. I kind of hesitant to put more logic in this holistic method.